src/api/streams/binary-stream-manager.ts

```diff
@@ -0,0 +1,248 @@
+      buffer.shift() // Remove oldest frame
+    }
+
+    // Send to all connected clients
+    const clients = this.clients.get(streamId)
+    if (!clients || clients.size === 0) {
+      return // No clients, buffered for later
+    }
+
+    for (const client of clients) {
+      this.sendToClient(client, data)
+    }
+  }
+
+  /**
+   * Add a WebSocket client to a stream
+   *
+   * @param streamId - Stream identifier
+   * @param ws - WebSocket connection
+   * @param principal - Security principal (authenticated user)
+   */
+  addClient(streamId: string, ws: WebSocket, principal: any): void {
+    debug(`Adding client to stream: ${streamId}`)
+    console.log(`[binaryStreamManager] addClient: streamId="${streamId}"`)
+
+    const client: StreamClient = {
+      streamId,
+      ws,
+      principal,
+      backpressureDropCount: 0,
+      connectedAt: Date.now()
+    }
+
+    // Add to clients map
+    let clients = this.clients.get(streamId)
+    if (!clients) {
+      clients = new Set()
+      this.clients.set(streamId, clients)
+    }
+    clients.add(client)
+
+    // Note: We intentionally do NOT send buffered frames to new clients.
+    // For high-frequency streams like radar, sending 100 buffered frames
+    // at once overwhelms the client's WebSocket buffer, triggering
+    // backpressure disconnection. Fresh data will arrive immediately anyway.
+
+    debug(`Stream ${streamId} now has ${clients.size} client(s)`)
+  }
+
+  /**
+   * Remove a WebSocket client from a stream
+   *
+   * @param streamId - Stream identifier
+   * @param ws - WebSocket connection to remove
+   */
+  removeClient(streamId: string, ws: WebSocket): void {
+    const clients = this.clients.get(streamId)
+    if (!clients) {
+      return
+    }
+
+    // Find and remove client
+    for (const client of clients) {
+      if (client.ws === ws) {
+        clients.delete(client)
+        debug(`Removed client from stream: ${streamId}`)
+        console.log(
+          `[binaryStreamManager] removeClient: streamId="${streamId}", remaining=${clients.size}`
+        )
+        break
+      }
+    }
+
+    // Clean up empty client sets
+    if (clients.size === 0) {
+      this.clients.delete(streamId)
+      debug(`Stream ${streamId} has no more clients`)
+    }
+  }
+
+  /**
+   * Clean up all clients and buffers for a stream
+   * Called when plugin stops
+   *
+   * @param streamId - Stream identifier
+   */
+  cleanupStream(streamId: string): void {
+    const clients = this.clients.get(streamId)
+    if (clients) {
+      for (const client of clients) {
+        try {
+          client.ws.close(1001, 'Stream ended')
+        } catch (error) {
+          debug(`Error closing client WebSocket: ${error}`)
+        }
+      }
+      this.clients.delete(streamId)
+    }
+
+    this.buffers.delete(streamId)
+    debug(`Cleaned up stream: ${streamId}`)
+  }
+
+  /**
+   * Send binary data to a specific client with backpressure handling
```

Comment:
A nitpick, but to me this is not proper backpressure handling, but _slow consumer disconnect_ or something like that.